core/checksums/src/main/java/software/amazon/awssdk/checksums/internal/DigestAlgorithm.java

```diff
@@ -27,41 +31,92 @@ public enum DigestAlgorithm {
+
+    private MessageDigest newDigest() {
+        try {
+            return MessageDigest.getInstance(algorithmName);
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException("Unable to fetch message digest instance for Algorithm "
+                                       + algorithmName + ": " + e.getMessage(), e);
+        }
     }
 
-    private static class DigestThreadLocal extends ThreadLocal<MessageDigest> {
-        private final String algorithmName;
+    public final class CloseableMessageDigest implements SdkAutoCloseable, Cloneable {
+        private final Supplier<MessageDigest> closedDigest = () -> {
+            throw new IllegalStateException("This message digest is closed.");
+        };
+        private Supplier<MessageDigest> digest;
+        private byte[] messageDigest;
+
+        private CloseableMessageDigest(MessageDigest digest) {
+            this.digest = () -> digest;
+        }
+
+        /**
+         * Retrieve the message digest instance.
+         */
+        public MessageDigest messageDigest() {
+            return digest.get();
+        }
 
-        DigestThreadLocal(String algorithmName) {
-            this.algorithmName = algorithmName;
+        /**
+         * Retrieve the message digest bytes. This will close the message digest when invoked. This is because the underlying
+         * message digest is reset on read, and we'd rather fail future interactions with the digest than act on the wrong data.
+         */
+        public byte[] digest() {
+            if (messageDigest != null) {
+                return messageDigest;
+            }
+            messageDigest = messageDigest().digest();
+            return messageDigest;
+        }
+
+        /**
+         * Release this message digest back to the cache. Once released, you must not use the digest anymore.
+         */
+        @Override
+        public void close() {
+            if (digest == closedDigest) {
+                return;
+            }
+
+            // Avoid over-caching after large traffic bursts. The maximum chosen here is arbitrary. It's also not strictly
+            // enforced, since these statements aren't synchronized.
+            if (digestCache.size() <= MAX_CACHED_DIGESTS) {
+                digestCache.addFirst(digest.get());
+            }
+            digest = closedDigest;
         }
 
         @Override
-        protected MessageDigest initialValue() {
+        public CloseableMessageDigest clone() {
             try {
-                return MessageDigest.getInstance(algorithmName);
-            } catch (NoSuchAlgorithmException e) {
-                throw new RuntimeException("Unable to fetch message digest instance for Algorithm "
-                                           + algorithmName + ": " + e.getMessage(), e);
+                return new CloseableMessageDigest((MessageDigest) digest.get().clone());
+            } catch (CloneNotSupportedException e) { // should never occur
```

Comment:
Any time a comment says "should never occur", it seems to happen.  Why can't this method declare that it throws CloneNotSupportedException instead?

Reply:
`CloneNotSupportedException` is a checked exception. Because the SDKs don't throw checked exceptions, the callers would need to wrap it in an unchecked exception themselves.

Reply:
I've improved the exception message if this scenario does happen.